### PR TITLE
Add Array.find_if()/rfind_if() methods

### LINKED ISSUE #6006: Error instancing scene as editor tool: Condition ' signal_map.has(p_signal.name) ' is true
Tested on Godot 2.1 rc1 Windows 10 64 bits

I get this error when I launch this test project:
[PluginInstanceSignalError.zip](https://github.com/godotengine/godot/files/395117/PluginInstanceSignalError.zip)
This project contains a plugin that instances a custom control scene. This scene contains a HButtonArray. If I remove it, i don't get the error.

It's possibly related to https://github.com/godotengine/godot/issues/6005.

### LINKED ISSUE #6006: Add a find_first function to Array functions
### Describe the project you are working on

Applies to nearly all of my different game projects.

### Describe the problem or limitation you are having in your project

Very often I have an array of objects/nodes which I need to iterate through in search of the first one which satisfies a more complex condition, than just its own value.

For example:

```
func find_faction(faction_id):
	for node in $Factions.get_children():
		if node.data.id == faction_id:
			return node
	return null
```

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It could be implemented by means of a function in the new Array functions, similar to any, all, map etc.
I notice that 'any' performs almost the same but it returns true or false. The find_first function would instead return the matching object, or null.

`Variant find_first(Callable method) const`

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The above example could be rewritten like this:

```
func find_faction(faction_id):
	var arr = $Factions.get_children()
	return arr.find_first(func(node): return node.data.id == faction_id)
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It can be worked around, as I'm currently doing. However, I think it will be used a lot and I also believe that it can be easily implemented by just using the code implementation for 'any' and tweak it to return the object instead of true, or null if no object is found.

### Is there a reason why this should be core and not an add-on in the asset library?

It would be a function which is part of the Array functions, thus core.

### LINKED ISSUE #6889: Add more array methods: first_custom, last_custom, count_custom
### Describe the project you are working on

This is not unique to my project, but one project I am working on requires a lot of array traversals searching for elements that satisfy specific conditions. I feel there is a gap in Array functionality (because sort_custom already exists), and this would plug it,

### Describe the problem or limitation you are having in your project

There is no built-in methods to search for the first, last, and to count elements that satisfy custom conditions implemented in a Callable, the way you can filter or sort_custom.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add the following methods:
`find_custom(fn:Callable)` will return the index of the first element of an array where the callable fn will return true.
`rfind_custom(fn:Callable)` will return the index of the last element of an array where the callable fn will return true.
`count_custom(fn:Callable)` will return the count of all elements of the array where the callable fn will return true.

I attempted to name the functions to fit the naming style of the rest of the functions in the Array class. However, I do not like it- I would rather name it something like `find_by()`. But my gripes with the Array function naming conventions are beyond the scope of this proposal.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Here is a sketch implementation I have already created:
```cpp
int Array::find_custom(const Callable& p_callable, int p_from = 0) const {
	if (_p->array.size() == 0) {
		return -1;
	}

	if (p_from < 0 || size() == 0) {
		return -1;
	}

	const Variant* argptrs[1];

	for (int i = p_from; i < size(); i++) {
		argptrs[0] = &get(i);

		Variant result;
		Callable::CallError ce;

		p_callable.callp(argptrs, 1, result, ce);
		if (ce.error != Callable::CallError::CALL_OK) {
			ERR_FAIL_V_MSG(-1, "Error calling method from 'first_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
		}
		// Return as soon as possible
		if (result.operator bool()) {
			return i;
		}
	}

	return -1;
}

int Array::rfind_custom(const Callable& p_callable, int p_from = -1) const {
	if (_p->array.size() == 0) {
		return -1;
	}

	if (p_from < 0) {
		// Relative offset from the end
		p_from = _p->array.size() + p_from;
	}
	if (p_from < 0 || p_from >= _p->array.size()) {
		// Limit to array boundaries
		p_from = _p->array.size() - 1;
	}

	const Variant* argptrs[1];

	for (int i = p_from; i >= 0; i--) {
		argptrs[0] = &get(i);

		Variant result;
		Callable::CallError ce;

		p_callable.callp(argptrs, 1, result, ce);
		if (ce.error != Callable::CallError::CALL_OK) {
			ERR_FAIL_V_MSG(-1, "Error calling method from 'last_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
		}
		// Return as soon as possible
		if (result.operator bool()) {
			return i;
		}
	}

	return -1;
}

int Array::count_custom(const Callable& p_callable) const {
	int count = 0;
	const Variant* argptrs[1];

	for (int i = 0; i < size(); i++) {
		argptrs[0] = &get(i);

		Variant result;
		Callable::CallError ce;

		p_callable.callp(argptrs, 1, result, ce);
		if (ce.error != Callable::CallError::CALL_OK) {
			ERR_FAIL_V_MSG(0, "Error calling method from 'count_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
		}
		// Increase count if passes
		if (result.operator bool()) {
			count++;
		}
	}

	return count;
}
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, one can simply create a loop to accomplish the same task. But it must be done every time, and could be solved by patching this gap in GDScript.

### Is there a reason why this should be core and not an add-on in the asset library?

This adds core Array functionality.

### PR DESCRIPTION
**Pull request by maximkulkin**
Add methods to Array to find elements based on predicate.

Fixes godotengine/godot-proposals#6006

* Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/6889*

### CODE REVIEW

**Comment by filkata** [comment]
Would love to see this merged as it is something that will be useful for me.

**Comment by AThousandShips** [comment]
This must have been missed along with the proposal when this was opened:
* https://github.com/godotengine/godot/pull/80208

File: core/variant/array.cpp [hunk 1]
```diff
@@ -41,6 +41,8 @@
 #include "core/variant/dictionary.h"
 #include "core/variant/variant.h"
 
+#include <algorithm>
```
Comment (dalexeev) [inline core/variant/array.cpp:44]: This doesn't seem to be used.

File: tests/core/variant/test_array.h [hunk 2]
```diff
@@ -129,6 +129,170 @@ TEST_CASE("[Array] has() and count()") {
...
+template <typename F>
+Callable predicate(F &&f) { return callable_lambda<bool, const Variant &>(f); }
+
+bool operator>(const Variant &l, const Variant &r) { return l != r && !(l < r); }
+
+inline bool operator<(const Variant &v, int x) { return v < Variant(x); }
+inline bool operator>(const Variant &v, int x) { return v > Variant(x); }
+
+TEST_CASE("[Array] find_if()") {
+	Callable less_than_3 = predicate([](auto &v) { return v < 3; });
```
Comment (dalexeev) [inline tests/core/variant/test_array.h:224]: The `auto` keyword is not allowed, see [docs](https://docs.godotengine.org/en/stable/contributing/development/cpp_usage_guidelines.html#auto-keyword).

**Review by dalexeev — COMMENTED** [review]
These methods should be named `find_custom()` and `rfind_custom()`, for consistency with the existing API.

**Comment by AThousandShips** [comment]
The naming should match the rest of the functions taking a callable, but no need to do any additional work before a decision on which will get merged is made

File: core/variant/array.h [hunk 3]
```diff
@@ -89,6 +89,8 @@ class Array {
 
 	int find(const Variant &p_value, int p_from = 0) const;
 	int rfind(const Variant &p_value, int p_from = -1) const;
+	int find_if(const Callable &p_func, int p_from = 0) const;
```
Comment (AThousandShips) [inline core/variant/array.h:92]:
```suggestion
	int find_if(const Callable &p_callable, int p_from = 0) const;
```
Match the other ones taking a callable, keep it `func` in the bindings

File: doc/classes/Array.xml [hunk 4]
```diff
@@ -309,6 +309,14 @@
 				Searches the array for a value and returns its index or [code]-1[/code] if not found. Optionally, the initial search index can be passed.
 			</description>
 		</method>
+		<method name="find_if" qualifiers="const">
+			<return type="int" />
+			<param index="0" name="func" type="Callable" />
+			<param index="1" name="from" type="int" default="0" />
+			<description>
+				Searches the array for the first element for which given function returns [code]true[/code] and returns its index or [code]-1[/code] if not found. Optionally, the initial search index can be passed.
```
Comment (AThousandShips) [inline doc/classes/Array.xml:317]:
```suggestion
				Searches the array for the first element for which the provided [Callable] returns [code]true[/code] and returns its index or [code]-1[/code] if not found. Optionally, the initial search index can be passed.
```
Same below

File: tests/core/variant/test_array.h [hunk 5]
```diff
@@ -129,6 +129,170 @@ TEST_CASE("[Array] has() and count()") {
 	CHECK(arr.count(2) == 0);
 }
 
+TEST_CASE("[Array] find()") {
+	Array arr;
+	// Find on empty array returns nothing
```
Comment (AThousandShips) [inline tests/core/variant/test_array.h:134]:
```suggestion
	// Find on empty array returns nothing.
```
Per the [comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide). For all comments in your addition

File: tests/core/variant/test_array.h (hunk 2 above)
Comment (AThousandShips) [reply tests/core/variant/test_array.h:224]: Also let's [avoid lambas](https://docs.godotengine.org/en/stable/contributing/development/cpp_usage_guidelines.html#lambdas) here all together

File: core/variant/array.cpp [hunk 6]
```diff
@@ -353,6 +355,30 @@ int Array::find(const Variant &p_value, int p_from) const {
...
+
+	const Variant *argptrs[1];
+	for (int i = p_from; i < size(); i++) {
+		argptrs[0] = &get(i);
+
+		Variant result;
+		Callable::CallError ce;
+		p_func.callp(argptrs, 1, result, ce);
+		if (ce.error != Callable::CallError::CALL_OK) {
+			ERR_FAIL_V_MSG(-1, "Error calling method from 'find_if': " + Variant::get_callable_error_text(p_func, argptrs, 1, ce));
```
Comment (AThousandShips) [inline core/variant/array.cpp:371]:
```suggestion
			ERR_FAIL_V_MSG(-1, "Error calling method from 'find_if': " + Variant::get_callable_error_text(p_func, argptrs, 1, ce) + ".");
```

File: core/variant/array.cpp [hunk 7]
```diff
@@ -378,6 +404,34 @@ int Array::rfind(const Variant &p_value, int p_from) const {
...
+
+	const Variant *argptrs[1];
+	for (int i = p_from; i >= 0; i--) {
+		argptrs[0] = &get(i);
+
+		Variant result;
+		Callable::CallError ce;
+		p_func.callp(argptrs, 1, result, ce);
+		if (ce.error != Callable::CallError::CALL_OK) {
+			ERR_FAIL_V_MSG(-1, "Error calling method from 'rfind_if': " + Variant::get_callable_error_text(p_func, argptrs, 1, ce));
```
Comment (AThousandShips) [inline core/variant/array.cpp:424]:
```suggestion
			ERR_FAIL_V_MSG(-1, "Error calling method from 'rfind_if': " + Variant::get_callable_error_text(p_func, argptrs, 1, ce) + ".");
```

File: core/variant/array.cpp [hunk 8]
```diff
@@ -378,6 +404,34 @@ int Array::rfind(const Variant &p_value, int p_from) const {
 	return -1;
 }
 
+int Array::rfind_if(const Callable &p_func, int p_from) const {
+	if (p_from < 0) {
+		// Relative offset from the end
```
Comment (AThousandShips) [inline core/variant/array.cpp:409]:
```suggestion
		// Relative offset from the end.
```

**Review by AThousandShips — COMMENTED** [review]
Again, no need to make any work until a decision has been made

File: tests/core/variant/test_array.h [hunk 9]
```diff
@@ -129,6 +129,170 @@ TEST_CASE("[Array] has() and count()") {
 	CHECK(arr.count(2) == 0);
 }
 
+TEST_CASE("[Array] find()") {
```
Comment (AThousandShips) [inline tests/core/variant/test_array.h:132]: The tests for `find` and `rfind` belong in a separate PR

**Comment by AThousandShips** [comment]
Reviewers, see also:
* https://github.com/godotengine/godot/pull/70727
* https://github.com/godotengine/godot/pull/80208

**Comment by maximkulkin** [comment]
> These methods should be named `find_custom()` and `rfind_custom()`, for consistency with the existing API.

I disagree. First, there are only two existing methods that take a predicate and have "_custom" suffix - "sort_custom" and "bsearch_custom" - for which other alternatives are not suitable, which hardly defines a standard for a naming convention. While there are other, more intuitive and accepted in other languages alternatives. I chose "_if" suffix ("find_if"/"rfind_if") as this API is implemented and to be used in C++ also, and in C++ those are established names for such algorithms. "find_custom" is a non-intuitive name and requires knowledge of other Array APIs to guess the similarities, and those are not the most common operations to even be aware of (especially for less experienced programmers).

**Comment by AThousandShips** [comment]
There's plenty of other APIs which all have their own style, but there's no use of `*_if` syntax anywhere in the engine, so one is established and the other one is. This is not the c++ standard library

It is up to you but I would prefer the syntax to match and not introduce new syntax

**Comment by maximkulkin** [comment]
There's no use of "*_if", because there's no API that does search by predicate. The only API that comes close is Array.filter, but it's basic form already does predicates. I have checked the whole code base and haven't found any similar "custom" API. All I see is "custom" being used as a go to term to describe APIs that set. So, in different APIs word "custom" means different things (either do it differently or use override) and I would argue this "syntax" as you call it is a good idea. A couple of examples:
```c++
Node *MultiplayerSpawner::instantiate_custom(const Variant &p_data);
Error FileAccessMemory::open_custom(const uint8_t *p_data, uint64_t p_len);
void MaterialStorage::sampler_rd_configure_custom(float p_mipmap_bias);
void SurfaceTool::set_custom(int p_channel_index, const Color &p_custom);
```

**Comment by AThousandShips** [comment]
But for `Array` `*_custom` means "takes a custom callable to do the action", be that "custom predicate" or "custom comparator"

It's either way, you pick what you prefer, but two reviewers have suggested this so it's something to consider for approval

**Comment by akien-mga** [comment]
Superseded by #95449. Thanks for the contribution!
